tracing/trace_keyspace_helper.cc

```diff
@@ -289,10 +289,10 @@ cql3::query_options trace_keyspace_helper::make_slow_query_mutation_data(const o
     const session_record& record = session_records.session_rec;
     auto millis_since_epoch = std::chrono::duration_cast<std::chrono::milliseconds>(record.started_at.time_since_epoch()).count();
 
+    sstring query_str;
     // query command is stored on a parameters map with a 'query' key
-    auto query_str_it = record.parameters.find("query");
-    if (query_str_it == record.parameters.end()) {
-        throw std::logic_error("No \"query\" parameter set for a session requesting a slow_query_log record");
+    if (auto query_str_it = record.parameters.find("query"); query_str_it != record.parameters.end()) {
+        query_str = query_str_it->second;
```

Comment:
maybe just sstring query_str = record.parameters["query"]; ?

Reply:
I didn't try this, but it's a `const map`..? And `operator[]` always creates the key..?

Reply:
If it's const then it won't work

Reply:
This check makes sense for non-batch queries.
If it's not a batch query - then not having a `query` string is a bug.
If it's a batch query then we should probably see what Cassandra does in this case and do the same.

For instance I see that node_slow_log has a set<text> for tables names which is clearly defined with a batch in mind.

Reply:
Even if not having a `query` string is a bug, tracing this entry still makes sense in my opinion. `std::logic_error("No \"query\" parameter set for a session requesting a slow_query_log record")` absolutely lacks context by itself, so it's much better to still trace the queryless request, and perhaps also log a warning which says that the `query` parameter was missing.

Reply:
@psarna I guess you would agree that tracing a query without knowing what the query was is absolutely pointless. 

Note that the node_slow_log has a special column for a query string: "command".
The whole point of having this table (`node_slow_log`) is to be able to see which query has been slow by only looking at it and not at anything else.

When you think about the implementation for this GH issue you should imagine that there is no extra tracing data except for what's written in node_slow_log. That's how it is in Cassandra and that's how it should have been all alone.

I want to implement it properly at some point: https://github.com/scylladb/scylla/issues/2572

Reply:
I agree that tracing a query without knowing what it is is pointless, but it's not the case here - after this patch, we'll just sometimes trace a query which doesn't have a `query` entry in its parameters, which is not the same as not knowing what it is. For instance, with batches we know perfectly fine what the queries are, because this information can be extracted from other fields. I think that throwing this generic `std::logic_error("No \"query\" parameter set for a session requesting a slow_query_log record"` gives us very little context on what actually happened, while a trace entry without the `query` field provides every bit of information we need. Throwing this exception in this particular case is literally throwing away context which could help us pinpoint the issue, so I think that this patch is a clear improvement.

Reply:
@psarna You have missed my point: this issue (and PR) is not about `system_traces.sessions` table which has (today) the query in the parameters.
This issue is about the `system_traces.node_slow_log` which can only have the query string in the `command` column and it is intended to be used stand alone.

The fact that we today have entries in `sessions` and `events` tables for each entry in `node_slow_log` is design bug and has to be fixed ASAP.

Because of that Node Slow Log is an almost-dead feature because it's intended to be always-on. But because of that design bug it can't be because it introduces a heavy overhead via `sessions` and `events` tables.

Once we implement #2572 it will become operational and there won't be any entries is `sessions` and `events`.

Reply:
You've got me confused @vladzcloudius. First, you said we shouldn't concatenate batch queries and put them into a node_slow_log because these queries can be obtained in a different way and now you say that the can't. If that's true and Node Slow Log is meant to be a standalone feature then the initial version of the PR was correct to use a concatenation of all batch elements as a query field.

Reply:
Well, I'm a person, @haaawk . I make mistakes too.
The best way to move forward is to look what C* does.
Printing the whole batch query which may be VERY long doesn't sound like an obvious solution to me either.

I'd suggest to use some length limited string - similar to what we do for prepared statements values.

However, again, there is no need to invent a bicycle here - better follow C* on this.

Reply:
Limiting the size sounds like a good idea. @jul-stas please check what C* is doing and let us know. Thank you!